### PR TITLE
GUACAMOLE-36: Add support for tracking password history

### PR DESCRIPTION
**Pull request by mike-jumper**
This change adds optional tracking and past passwords (hashes+salts only), and adds policy options which prevent password reuse.

### CODE REVIEW

File: extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/apache/guacamole/auth/jdbc/security/PasswordPolicy.java [hunk 1]
```diff
@@ -73,6 +73,20 @@
     int getMaximumAge() throws GuacamoleException;
 
     /**
+     * Returns the number of previous passwords remembered for each user. If
+     * greater than zero, users will be prohibited from reusing their past
+     * passwords.
```
Comment (jmuehlner) [inline extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/apache/guacamole/auth/jdbc/security/PasswordPolicy.java]: They'll be prohibited from reusing passwords in this history, right? This is a little vague and sounds like turning on this property ensures that a password can never be reused.

File: extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/apache/guacamole/auth/jdbc/user/ModeledUser.java [hunk 2]
```diff
@@ -189,10 +189,10 @@
     private String password = null;
 
     /**
-     * The time and date that this user's password was previously set (prior to
-     * being queried). If the user is new, this will be null.
+     * The data associated with this user's password at the time this user was
+     * queried. If the user is new, this will be null.
      */
-    private Timestamp previousPasswordDate = null;
```
Comment (jmuehlner) [inline extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/apache/guacamole/auth/jdbc/user/ModeledUser.java]: So... we're no longer using the "previousPasswordDate" column? Can it be removed?

File: extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-mysql/src/main/resources/org/apache/guacamole/auth/jdbc/user/PasswordRecordMapper.xml [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+-->
+
+<mapper namespace="org.apache.guacamole.auth.jdbc.user.PasswordRecordMapper" >
+
+    <!-- Result mapper for system permissions -->
```
Comment (jmuehlner) [inline extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-mysql/src/main/resources/org/apache/guacamole/auth/jdbc/user/PasswordRecordMapper.xml]: System permissions?

File: extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/apache/guacamole/auth/jdbc/security/PasswordPolicy.java (hunk 1 above)
Comment (mike-jumper) [inline extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/apache/guacamole/auth/jdbc/security/PasswordPolicy.java]: Yes, that's correct. I'll try to improve the wording.

File: extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-mysql/src/main/resources/org/apache/guacamole/auth/jdbc/user/PasswordRecordMapper.xml (hunk 3 above)
Comment (mike-jumper) [inline extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-mysql/src/main/resources/org/apache/guacamole/auth/jdbc/user/PasswordRecordMapper.xml]: Oops.

File: extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/apache/guacamole/auth/jdbc/user/ModeledUser.java (hunk 2 above)
Comment (jmuehlner) [inline extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/apache/guacamole/auth/jdbc/user/ModeledUser.java]: Ok

Comment (mike-jumper) [inline extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/apache/guacamole/auth/jdbc/user/ModeledUser.java]: What about simply `getPasswordRecord()`?

Comment (jmuehlner) [inline extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/apache/guacamole/auth/jdbc/user/ModeledUser.java]: Yeah, that part is a bit confusing as well. It took me a little bit to figure out what was going on. getPasswordRecord might be better, yeah.

**Comment by mike-jumper** [comment]
OK - I renamed `getPreviousPassword()` and `recordPreviousPassword()` to `getPasswordRecord()` and `recordPassword()` respectively.

**Comment by mike-jumper** [comment]
Rebased to fix the copy/paste issue with the password record mapping XML.

**Comment by jmuehlner** [comment]
Looks good now!

**Pull request merged**